### DISCUSSION TITLE
Approach for UserDefaults supported types but AppStorageKey not supported

### DISCUSSION THREAD

**Discussion by myihsan** [pointfreeco/swift-sharing#58] [Q&A]
For values that I control, I have defined a wrapper to handle them.
However, I want to retrieve a string array value set by `set(_ value: Any?, forKey defaultName: String)` that can only be retrieved by `stringArray(forKey:)` or `object(forKey:)` as far as I know.
So I created #56, which is closed due to the current stance, which also makes sense.

But how do we deal with these types?
It would be helpful to make the relative APIs @spi public so we can extend  `AppStorageKey` when we need to.

**Comment by mbrandonw** [comment]
Hi @myihsan, is there a reason you cannot use `.fileStorage` for more complex data types?

**Reply by myihsan** [reply]
Thanks for the suggestion! If I have control, I may use `.fileStorage`.
However, the problem is that I need to read and modify the values that others do as well.
For example, sync the language setting (a string array saved in `UserDefaults`) with Settings.app.

**Reply by mbrandonw** [reply]
> For example, sync the language setting (a string array saved in UserDefaults) with Settings.app

Can you explain this a bit more? What key are you wanting to modify in user defaults?

**Reply by myihsan** [reply]
The key is `AppleLanguages`.
I understand this is not a normal need😅

**Reply by mbrandonw** [reply]
I'm not sure it's legitimate to change that key in user defaults. See [this](https://forums.developer.apple.com/forums/thread/718512?answerId=733680022#733680022).

**Reply by myihsan** [reply]
I know. I only use it to sync the language setting with Settings.app and switch language by `Environment` since there is a requirement to do it without relaunching the app.

So it would be great to have the ability to extend AppStorageKey even by @spi public.

**Reply by myihsan** [reply]
https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/UserDefaults/AboutPreferenceDomains/AboutPreferenceDomains.html
> If you want to change the value of a preference in the global domain, write that same preference to the application domain with the new value.

According to this document, it seems legitimate for my use.
